prellblock-client/src/main.rs

```diff
@@ -130,6 +132,69 @@ async fn main_benchmark(identity: String, turi_address: Address, cmd: cmd::Bench
     }
 }
 
+async fn main_csv_replay(mut client: Client, cmd: cmd::CsvReplay) {
+    let cmd::CsvReplay {
+        csv_file,
+    } = cmd;
+    let mut reader = Reader::from_path(csv_file)
+                             .expect("Cannot open CSV file.");
+
+    let mut headers: Vec<String> = Vec::new();
+
+    // collect column names
+    for header in reader
+                  .headers()
+                  .expect("Error reading headers from CSV files." )
+                  .iter()
+                  .skip(1) {
+        headers.push(header.to_string());
+    }
+
+    let mut next_release = Instant::now();
+    let mut prev_timestamp: Option<f64> = None;
+
+    for result in reader.records() {
+        let row = result.expect("Error reading row from CSV file.");
+        let mut row_iter = row.iter();
+
+        let current_timestamp: f64 = row_iter
+                                  .next()
+                                  .expect("Error reading first column.")
+                                  .parse()
+                                  .expect("Error parsing timestamp.");
+
+        if !prev_timestamp.is_none() {
+
+            // csv data must be sorted by timestamp
+            assert!(current_timestamp >= prev_timestamp.unwrap());
+
+            next_release += Duration::from_secs_f64(
+                current_timestamp - prev_timestamp.unwrap()
+            );
+            if next_release < Instant::now() {
+                log::warn!("System cannot keep up replaying data fast \
+                            enough. {} seconds behind.",
+                            (Instant::now() - next_release).as_secs_f64())
+            } else {
+                // sorry but busy waiting is most precise and wait times
+                // are expected to be just a few seconds
+                while Instant::now() < next_release {}
```

Comment:
Did you try out `tokio::time::sleep_util(next_release).await`? I think prellblock operates single-threaded (using an event loop) and thus busy-waiting might have unexpected effects -- including the mentioned performance issue?

https://docs.rs/tokio/1.8.2/tokio/time/fn.sleep_until.html

Reply:
Thanks for the comment.
No, I didn't try sleeps at all, since their precision is always a bit of a lottery. When waiting to fire the next values, the client (process) should have nothing else to do – as far as I understand. :)
Instead of waiting for each `send_key_value` to return, I tried to fire all `send_key_value` asynchronously and then waiting for one after the other afterwards – this led to no noticeable performance gain.
The Prellblock RPUs are four separate processes, of course.
I can investigate if sleeps would help out of curiosity.

Reply:
kk -- I missed that this happens in the client process. Nevermind then

Reply:
Just did another quick improvised measurement: submitting a row of our data (project-internal, approx. 75 values per row) takes approx. .8 s. Rows would need to be submitted approx. every .5 s. (i.e., currently, the code path with the busy wait loop is never activated).
Nevertheless, it would definitively be an enhancement to sleep until – e.g. – 1 s before the next release and busy wait the last second. But oh well… :)